### ISSUE TITLE
[macOS-arm64] System.Net.Http.Functional.Tests failures on CoreCLR

### ISSUE THREAD

**Issue by sdmaclea** [dotnet/runtime#49109]
See https://dev.azure.com/dnceng/public/_build/results?buildId=1021947

```console
----- start Wed Mar 3 17:36:19 PST 2021 =============== To repro directly: =====================================================
pushd .
/tmp/helix/working/C24F0A5F/p/dotnet exec --runtimeconfig System.Net.Http.Functional.Tests.runtimeconfig.json --depsfile System.Net.Http.Functional.Tests.deps.json xunit.console.dll System.Net.Http.Functional.Tests.dll -xml testResults.xml -nologo -nocolor -notrait category=IgnoreForCI -notrait category=failing 
popd
===========================================================================================================
/private/tmp/helix/working/C24F0A5F/w/9A5308C9/e /private/tmp/helix/working/C24F0A5F/w/9A5308C9/e
  Discovering: System.Net.Http.Functional.Tests (method display = ClassAndMethod, method display options = None)
  Discovered:  System.Net.Http.Functional.Tests (found 1205 of 1415 test cases)
  Starting:    System.Net.Http.Functional.Tests (parallel test collections = on, max threads = 8)
    ...
    ...
    ...
System.Net.Http.Functional.Tests.SocketsHttpHandler_HttpClientHandler_Authentication_Test.Proxy_DomainJoinedProxyServerUsesKerberos_Success [SKIP]
      Condition(s) not met: "IsDomainJoinedServerAvailable"
Unhandled exception. System.NullReferenceException: Object reference not set to an instance of an object.
   at System.SZArrayHelper.get_Count[T]() in /_/src/coreclr/System.Private.CoreLib/src/System/Array.CoreCLR.cs:line 539
   at System.Collections.ObjectModel.ReadOnlyCollection`1.get_Count() in /_/src/libraries/System.Private.CoreLib/src/System/Collections/ObjectModel/ReadOnlyCollection.cs:line 26
   at System.Net.Http.Functional.Tests.TelemetryTest.ValidateStartEventPayload(EventWrittenEventArgs startEvent) in /_/src/libraries/System.Net.Http/tests/FunctionalTests/TelemetryTest.cs:line 413
   at System.Net.Http.Functional.Tests.TelemetryTest.<>c.<<EventSource_SuccessfulRequest_LogsStartStop>b__3_0>d.MoveNext() in /_/src/libraries/System.Net.Http/tests/FunctionalTests/TelemetryTest.cs:line 177
--- End of stack trace from previous location ---
   at Microsoft.DotNet.RemoteExecutor.Program.Main(String[] args) in /_/src/Microsoft.DotNet.RemoteExecutor/src/Program.cs:line 65
--- End of stack trace from previous location ---
   at Microsoft.DotNet.RemoteExecutor.Program.Main(String[] args) in /_/src/Microsoft.DotNet.RemoteExecutor/src/Program.cs:line 94
    ...
    ...
    ...
System.Net.Http.Functional.Tests.TelemetryTest_Http20.EventSource_SuccessfulRequest_LogsStartStop(testMethod: "GetStringAsync") [FAIL]
      Microsoft.DotNet.RemoteExecutor.RemoteExecutionException : Remote process failed with an unhandled exception.
      Stack Trace:
        
        Child exception:
          System.NullReferenceException: Object reference not set to an instance of an object.
        /_/src/coreclr/System.Private.CoreLib/src/System/Array.CoreCLR.cs(539,0): at System.SZArrayHelper.get_Count[T]()
        /_/src/libraries/System.Private.CoreLib/src/System/Collections/ObjectModel/ReadOnlyCollection.cs(26,0): at System.Collections.ObjectModel.ReadOnlyCollection`1.get_Count()
        /_/src/libraries/System.Net.Http/tests/FunctionalTests/TelemetryTest.cs(413,0): at System.Net.Http.Functional.Tests.TelemetryTest.ValidateStartEventPayload(EventWrittenEventArgs startEvent)
        /_/src/libraries/System.Net.Http/tests/FunctionalTests/TelemetryTest.cs(177,0): at System.Net.Http.Functional.Tests.TelemetryTest.<>c.<<EventSource_SuccessfulRequest_LogsStartStop>b__3_0>d.MoveNext()
        --- End of stack trace from previous location ---
        /_/src/Microsoft.DotNet.RemoteExecutor/src/Program.cs(65,0): at Microsoft.DotNet.RemoteExecutor.Program.Main(String[] args)
        
        Child process:
          System.Net.Http.Functional.Tests, Version=6.0.0.0, Culture=neutral, PublicKeyToken=cc7b13ffcd2ddd51 System.Net.Http.Functional.Tests.TelemetryTest+<>c System.Threading.Tasks.Task <EventSource_SuccessfulRequest_LogsStartStop>b__3_0(System.String, System.String)
        
        Child arguments:
          2.0, GetStringAsync
        
    ...
    ...
    ...
=== TEST EXECUTION SUMMARY ===
   System.Net.Http.Functional.Tests  Total: 3544, Errors: 0, Failed: 1, Skipped: 94, Time: 41.185s
/private/tmp/helix/working/C24F0A5F/w/9A5308C9/e
----- end Wed Mar 3 17:37:01 PST 2021 ----- exit code 1 ----------------------------------------------------------
```

This looks like it is a CoreCLR runtime issue.

**Comment by karelz** [comment]
@sdmaclea is it deterministic or intermittent? If intermittent, how often does it happen?

@MihaZupan can you check the telemetry test? `System.Net.Http.Functional.Tests.TelemetryTest_Http20.EventSource_SuccessfulRequest_LogsStartStop`

**Comment by MihaZupan** [comment]
Judging from the trace, accesing `ReadOnlyCollection`'s `Count` property triggers a NRE.

---

The `IList` stored in it is an `object[]` generated by `EventSource`. This array can't be null (ctor throws).

Accesing `IList.Count` of an `object[]` is apparently done by https://github.com/dotnet/runtime/blob/f7395115bb809d27468d9944825a4bbd5f1acae1/src/coreclr/System.Private.CoreLib/src/System/Array.CoreCLR.cs#L510

A NRE is thrown by its `get_Count` method

https://github.com/dotnet/runtime/blob/f7395115bb809d27468d9944825a4bbd5f1acae1/src/coreclr/System.Private.CoreLib/src/System/Array.CoreCLR.cs#L535

So looks like a runtime issue

**Comment by sdmaclea** [comment]
> @sdmaclea is it deterministic or intermittent? If intermittent, how often does it happen?

It happened the first time the test was run on Apple Silicon.  It didn't occur on the rerun.

I haven't enable the libraries tests in CI (need to stabilize first), so there are no other runs.

It looks consistent with other intermittent runtime issues we are seeing.

**Comment by VincentBu** [comment]
Failed again in [runtime-libraries-coreclr outerloop 20210401.4](https://dev.azure.com/dnceng/public/_build/results?buildId=1067573&view=ms.vss-test-web.build-test-results-tab&runId=32839632&resultId=103184&paneView=debug)

Failed test:
```
net6.0-windows-Release-x64-CoreCLR_release-Windows.81.Amd64.Open
 -System.Net.Http.Functional.Tests.TelemetryTest_Http20.EventSource_SuccessfulRequest_LogsStartStop(testMethod: \"UnbufferedSendAsync\")
```

**Error message:**
```
Microsoft.DotNet.RemoteExecutor.RemoteExecutionException : Remote process failed with an unhandled exception.


Stack trace

Child exception:
  Xunit.Sdk.EqualException: Assert.Equal() Failure
Expected: 0
Actual:   1
   at System.Net.Http.Functional.Tests.TelemetryTest.VerifyEventCounters(ConcurrentQueue`1 events, Int32 requestCount, Boolean shouldHaveFailures, Int32 requestsLeftQueueVersion) in /_/src/libraries/System.Net.Http/tests/FunctionalTests/TelemetryTest.cs:line 514
   at System.Net.Http.Functional.Tests.TelemetryTest.<>c.<<EventSource_SuccessfulRequest_LogsStartStop>b__3_0>d.MoveNext() in /_/src/libraries/System.Net.Http/tests/FunctionalTests/TelemetryTest.cs:line 192
--- End of stack trace from previous location ---
   at Microsoft.DotNet.RemoteExecutor.Program.Main(String[] args) in /_/src/Microsoft.DotNet.RemoteExecutor/src/Program.cs:line 65

Child process:
  System.Net.Http.Functional.Tests, Version=6.0.0.0, Culture=neutral, PublicKeyToken=cc7b13ffcd2ddd51 System.Net.Http.Functional.Tests.TelemetryTest+<>c System.Threading.Tasks.Task <EventSource_SuccessfulRequest_LogsStartStop>b__3_0(System.String, System.String)

Child arguments:
  2.0, UnbufferedSendAsync
```

**Comment by wfurt** [comment]
May be different issue than originally reported. This maps to 
```c#
           Assert.Equal(0, currentRequests[^1]);
```

**Comment by sdmaclea** [comment]
This may be fixed by macOS 11.3 when it is released.

**Comment by wfurt** [comment]
Out of curiosity: what is different about 11.3 @sdmaclea?

**Comment by sdmaclea** [comment]
I don't happen to know.  I would guess Apple has fixed issues in the kernel or pthreads.

**Comment by sdmaclea** [comment]
CI likely fixed by dotnet/core-eng#12776

Feel free to reopen if it reoccurs now that the queues have been updated to macOS 11.3.
